File: Remoter-Mac/Sources/RemoterAgent/Server/WebSocketServer.swift
import Foundation
import Network
import CryptoKit
import Security

typealias MessageHandler    = (NWConnection, String) -> Void
typealias BinaryHandler     = (NWConnection, Data)   -> Void
typealias DisconnectHandler = (NWConnection)          -> Void

final class WebSocketServer {
    var onConnect:    ((NWConnection) -> Void)?
    var onText:       MessageHandler?
    var onBinary:     BinaryHandler?
    var onDisconnect: DisconnectHandler?

    /// When set, plain HTTP GET requests on this port are served from this directory.
    var webDir: URL?

    private var listener: NWListener?
    private let queue = DispatchQueue(label: "remoter.server", qos: .userInteractive)

    func start(port: UInt16) throws {
        let p = NWEndpoint.Port(rawValue: port)!
        // TLS (HTTPS/WSS) so browsers get a secure context and can use WebCodecs (H.264).
        // Falls back to plain TCP if the embedded cert is missing.
        let tlsParams = Self.makeTLSParameters()
        let params = tlsParams ?? .tcp
        let scheme = tlsParams != nil ? "wss/https" : "ws/http"
        ConnectionLogger.shared.logStep(sessionId: "tls", step: "listen_scheme", detail: scheme)
        let l = try NWListener(using: params, on: p)
        l.stateUpdateHandler = { state in
            switch state {
            case .ready:        print("[Server] Listening on :\(port) (\(scheme))")
            case .failed(let e): print("[Server] Failed: \(e)")
            default: break
            }
        }
        l.newConnectionHandler = { [weak self] conn in self?.accept(conn) }
        l.start(queue: queue)
        listener = l
    }

    func stop() {
        listener?.cancel()
        listener = nil
    }

    // MARK: - TLS

    private static func makeTLSParameters() -> NWParameters? {
        guard let identity = loadIdentity() else { return nil }
        let tlsOptions = NWProtocolTLS.Options()
        sec_protocol_options_set_local_identity(tlsOptions.securityProtocolOptions, identity)
        return NWParameters(tls: tlsOptions)
    }

    // Retain the identity for the process lifetime so NW Framework can access the
    // private key at any point during TLS handshakes without it being deallocated.
    private static var tlsIdentity: SecIdentity?

    /// Load TLS identity entirely in-process — no keychain involved.
    ///
    /// SecPKCS12Import without kSecImportExportKeychain returns a temporary SecIdentity
    /// whose private key lives in process memory. secd is never consulted, so macOS
    /// never shows a keychain authorization dialog regardless of how many clients connect.
    private static func loadIdentity() -> sec_identity_t? {
        guard let p12URL  = Bundle.main.url(forResource: "server", withExtension: "p12"),
              let p12Data = try? Data(contentsOf: p12URL) else {
            ConnectionLogger.shared.logStep(sessionId: "tls", step: "p12_not_found")
            return nil
        }

        let opts: [String: Any] = [kSecImportExportPassphrase as String: "remoter"]
        var items: CFArray?
        let status = SecPKCS12Import(p12Data as CFData, opts as CFDictionary, &items)
        guard status == errSecSuccess,
              let arr  = items as? [[String: Any]],
              let first = arr.first,
              let id   = first[kSecImportItemIdentity as String] as! SecIdentity? else {
            ConnectionLogger.shared.logStep(sessionId: "tls", step: "p12_import_failed",
                detail: "status=\(status)")
            return nil
        }

        tlsIdentity = id
        ConnectionLogger.shared.logStep(sessionId: "tls", step: "identity_ok")
        return sec_identity_create(id)
    }

    func sendText(_ text: String, to conn: NWConnection) {
        guard let data = text.data(using: .utf8) else { return }
        sendFrame(opcode: 0x01, payload: data, to: conn)
    }

    func sendBinary(_ data: Data, to conn: NWConnection) {
        sendFrame(opcode: 0x02, payload: data, to: conn)
    }

    /// Video-frame variant: fires onSent when TCP has accepted the data, giving
    /// the caller real backpressure so frames are never queued ahead of the
    /// network's actual capacity.
    func sendBinaryVideo(_ data: Data, to conn: NWConnection, onSent: @escaping () -> Void) {
        var frame = Data()
        frame.append(0x82)                          // FIN + binary opcode
        let len = data.count
        if len < 126 {
            frame.append(UInt8(len))
        } else if len < 65536 {
            frame.append(126)
            frame.append(UInt8((len >> 8) & 0xFF))
            frame.append(UInt8( len       & 0xFF))
        } else {
            frame.append(127)
            for i in (0..<8).reversed() { frame.append(UInt8((len >> (i * 8)) & 0xFF)) }
        }
        frame.append(contentsOf: data)
        // One-shot guard: ensures onSent() fires exactly once even when both
        // the watchdog and contentProcessed callback race.
        let lock = NSLock()
        var completed = false
        let finish: (Bool) -> Void = { [weak self] cancel in
            lock.lock()
            let already = completed
            completed = true
            lock.unlock()
            guard !already else { return }
            if cancel {
                ConnectionLogger.shared.logStep(sessionId: "ws", step: "video_send_timeout")
                conn.cancel()
                // Immediately notify disconnect via server queue so the session's
                // encoder stops right away — avoids a flood of video_send_err while
                // the receive path slowly detects the cancel.
                self?.queue.async { [weak self] in self?.onDisconnect?(conn) }
            }
            onSent()
        }
        // Watchdog: if TCP flow-control stalls contentProcessed (client recv buffer full),
        // cancel the connection after 5s so the client can reconnect cleanly.
        DispatchQueue.global(qos: .utility).asyncAfter(deadline: .now() + 5) { finish(true) }
        conn.send(content: frame, completion: .contentProcessed { error in
            if let error {
                ConnectionLogger.shared.logStep(sessionId: "ws", step: "video_send_err",
                    detail: "\(error)")
            }
            finish(false)
        })
    }

    // MARK: - Accept + HTTP/WS dispatch

    private func accept(_ conn: NWConnection) {
        conn.start(queue: queue)
        conn.receive(minimumIncompleteLength: 1, maximumLength: 8192) { [weak self] data, _, _, error in
            guard let self, let data, !data.isEmpty, error == nil else { conn.cancel(); return }
            let request = String(data: data.prefix(4096), encoding: .ascii) ?? ""
            if let wsKey = Self.extractHeader(request, "Sec-WebSocket-Key") {
                self.upgradeWS(conn: conn, key: wsKey)
            } else {
                self.serveHTTP(request: request, conn: conn)
            }
        }
    }

    // MARK: - WebSocket upgrade + frame loop

    private func upgradeWS(conn: NWConnection, key: String) {
        let combined = key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
        let accept   = Data(Insecure.SHA1.hash(data: Data(combined.utf8))).base64EncodedString()
        let resp     = "HTTP/1.1 101 Switching Protocols\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Accept: \(accept)\r\n\r\n"
        conn.send(content: Data(resp.utf8), completion: .contentProcessed { [weak self] _ in
            self?.onConnect?(conn)        // 握手完成即通知，不等第一条消息
            self?.recvFrames(conn: conn, reader: WsFrameReader())
        })
    }

    private func recvFrames(conn: NWConnection, reader: WsFrameReader) {
        conn.receive(minimumIncompleteLength: 1, maximumLength: 65536) { [weak self] data, _, _, error in
            guard let self else { return }
            guard let data, !data.isEmpty, error == nil else {
                ConnectionLogger.shared.logStep(sessionId: "ws", step: "recv_disconnect",
                    detail: "err=\(String(describing: error)) emptyData=\(data?.isEmpty ?? true)")
                self.onDisconnect?(conn); return
            }
            var close = false
            reader.feed(data) { opcode, payload in
                switch opcode {
                case 0x01:
                    if let text = String(data: payload, encoding: .utf8) { self.onText?(conn, text) }
                case 0x02:
                    self.onBinary?(conn, payload)
                case 0x08:
                    close = true
                case 0x09: // ping → pong
                    self.sendFrame(opcode: 0x0A, payload: payload, to: conn)
                default: break
                }
            }
            if close {
                ConnectionLogger.shared.logStep(sessionId: "ws", step: "client_close_frame")
                self.onDisconnect?(conn); conn.cancel()
            }
            else { self.recvFrames(conn: conn, reader: reader) }
        }
    }

    private func sendFrame(opcode: UInt8, payload: Data, to conn: NWConnection) {
        var frame = Data()
        frame.append(0x80 | opcode)
        let len = payload.count
        if len < 126 {
            frame.append(UInt8(len))
        } else if len < 65536 {
            frame.append(126)
            frame.append(UInt8((len >> 8) & 0xFF))
            frame.append(UInt8( len       & 0xFF))
        } else {
            frame.append(127)
            for i in (0..<8).reversed() { frame.append(UInt8((len >> (i * 8)) & 0xFF)) }
        }
        frame.append(contentsOf: payload)
        conn.send(content: frame, completion: .idempotent)
    }

    // MARK: - HTTP file serving

    private func serveHTTP(request: String, conn: NWConnection) {
        guard let webDir else { conn.cancel(); return }
        let firstLine = request.components(separatedBy: "\r\n").first ?? ""
        let parts     = firstLine.components(separatedBy: " ")
        guard parts.count >= 2, parts[0] == "GET" else { conn.cancel(); return }

        let rawPath = parts[1].components(separatedBy: "?").first ?? "/"
        let rel     = rawPath == "/" ? "index.html" : String(rawPath.drop(while: { $0 == "/" }))
        var fileURL = webDir.appendingPathComponent(rel)
        if !FileManager.default.fileExists(atPath: fileURL.path) {
            fileURL = webDir.appendingPathComponent("index.html")
        }
        guard let body = try? Data(contentsOf: fileURL) else {
            let r = "HTTP/1.1 404 Not Found\r\nContent-Length: 0\r\n\r\n"
            conn.send(content: Data(r.utf8), completion: .contentProcessed { _ in conn.cancel() })
            return
        }
        let mime   = Self.mimeType(for: fileURL.pathExtension)
        let header = "HTTP/1.1 200 OK\r\nContent-Type: \(mime)\r\nContent-Length: \(body.count)\r\nCache-Control: no-cache\r\nConnection: close\r\n\r\n"
        var response = Data(header.utf8)
        response.append(body)
        conn.send(content: response, completion: .contentProcessed { _ in conn.cancel() })
    }

    private static func mimeType(for ext: String) -> String {
        switch ext.lowercased() {
        case "html":         return "text/html; charset=utf-8"
        case "js", "mjs":    return "application/javascript"
        case "css":          return "text/css"
        case "ico":          return "image/x-icon"
        case "png":          return "image/png"
        case "jpg", "jpeg":  return "image/jpeg"
        case "svg":          return "image/svg+xml"
        case "woff2":        return "font/woff2"
        case "woff":         return "font/woff"
        case "json":         return "application/json"
        default:             return "application/octet-stream"
        }
    }

    // MARK: - Helpers

    private static func extractHeader(_ request: String, _ name: String) -> String? {
        let prefix = name.lowercased() + ": "
        for line in request.components(separatedBy: "\r\n") {
            if line.lowercased().hasPrefix(prefix) {
                return String(line.dropFirst(prefix.count)).trimmingCharacters(in: .whitespaces)
            }
        }
        return nil
    }
}

// MARK: - WebSocket frame accumulator (handles fragmented TCP reads)

final class WsFrameReader {
    private var buf = Data()

    func feed(_ data: Data, handler: (UInt8, Data) -> Void) {
        buf.append(data)
        while tryParse(handler: handler) { }
    }

    private func tryParse(handler: (UInt8, Data) -> Void) -> Bool {
        guard buf.count >= 2 else { return false }
        let masked     = (buf[1] & 0x80) != 0
        var headerLen  = 2
        var payloadLen = Int(buf[1] & 0x7F)
        if payloadLen == 126 {
            guard buf.count >= 4 else { return false }
            payloadLen = (Int(buf[2]) << 8) | Int(buf[3])
            headerLen  = 4
        } else if payloadLen == 127 {
            guard buf.count >= 10 else { return false }
            payloadLen = 0
            for i in 2..<10 { payloadLen = (payloadLen << 8) | Int(buf[i]) }
            headerLen  = 10
        }
        let maskLen = masked ? 4 : 0
        let total   = headerLen + maskLen + payloadLen
        guard buf.count >= total else { return false }

        let opcode     = buf[0] & 0x0F
        let maskStart  = headerLen
        var payload    = Data(buf[(maskStart + maskLen) ..< total])
        if masked {
            for i in 0..<payload.count { payload[i] ^= buf[maskStart + (i % 4)] }
        }
        buf = Data(buf[total...])
        handler(opcode, payload)
        return true
    }
}

File: Remoter-Win/Server/WebSocketServer.cs
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;

namespace RemoterWin;

// Raw TCP WebSocket server — no HttpListener, no admin required.
// Handles the HTTP upgrade handshake and WebSocket framing manually.
sealed class WebSocketServer
{
    public Action<IWsConn>?         OnConnect;
    public Action<IWsConn, string>? OnText;
    public Action<IWsConn, byte[]>? OnBinary;
    public Action<IWsConn>?         OnDisconnect;

    private TcpListener? _listener;

    public void Start(int port)
    {
        _listener = new TcpListener(IPAddress.Any, port);
        _listener.Start();
        AppLog.Write($"[Server] Listening on :{port}");
        _ = AcceptLoopAsync();
    }

    public void Stop() => _listener?.Stop();

    private async Task AcceptLoopAsync()
    {
        while (true)
        {
            TcpClient client;
            try { client = await _listener!.AcceptTcpClientAsync(); }
            catch { break; }
            client.NoDelay = true;
            _ = HandleClientAsync(client);
        }
    }

    private async Task HandleClientAsync(TcpClient tcp)
    {
        var stream = tcp.GetStream();
        WsConn? conn = null;
        bool isHttpRequest = false;
        try
        {
            (conn, isHttpRequest) = await DoHandshakeAsync(stream, tcp);
            if (conn == null) 
            { 
                // If it was an HTTP request, wait a bit for data to be sent
                if (isHttpRequest)
                    await Task.Delay(500);
                tcp.Close(); 
                return; 
            }

            OnConnect?.Invoke(conn);  // fired before first message arrives
            await ReceiveLoopAsync(conn, stream);
        }
        catch (Exception ex) { AppLog.Write($"[WS] {conn?.RemoteAddr ?? "?"} error: {ex.GetType().Name}: {ex.Message}"); }
        finally
        {
            if (conn != null) OnDisconnect?.Invoke(conn);
            tcp.Close();
        }
    }

    // HTTP/WebSocket upgrade handshake
    // Returns (connection, isHttpRequest)
    private static async Task<(WsConn? conn, bool isHttp)> DoHandshakeAsync(NetworkStream stream, TcpClient tcp)
    {
        var buf  = new byte[4096];
        int read = 0;
        while (read < buf.Length)
        {
            int n = await stream.ReadAsync(buf.AsMemory(read));
            if (n == 0) return (null, false);
            read += n;
            var req = Encoding.ASCII.GetString(buf, 0, read);
            if (req.Contains("\r\n\r\n")) break;
        }

        var request = Encoding.ASCII.GetString(buf, 0, read);
        var key = ExtractHeader(request, "Sec-WebSocket-Key");
        if (key == null)
        {
            // Not a WebSocket upgrade — serve as HTTP GET if possible
            var served = await ServeFileAsync(stream, request);
            if (served)
            {
                return (null, true);
            }
            return (null, false);
        }

        var accept = Convert.ToBase64String(
            SHA1.HashData(Encoding.ASCII.GetBytes(key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"))
        );

        var response = Encoding.ASCII.GetBytes(
            "HTTP/1.1 101 Switching Protocols\r\n" +
            "Upgrade: websocket\r\n" +
            "Connection: Upgrade\r\n" +
            $"Sec-WebSocket-Accept: {accept}\r\n\r\n"
        );
        await stream.WriteAsync(response);
        return (new WsConn(stream, tcp), false);
    }

    private async Task ReceiveLoopAsync(WsConn conn, NetworkStream stream)
    {
        while (true)
        {
            var (opcode, payload) = await ReadFrameAsync(stream);
            switch (opcode)
            {
                case 0x1: // text
                    OnText?.Invoke(conn, Encoding.UTF8.GetString(payload));
                    break;
                case 0x2: // binary
                    OnBinary?.Invoke(conn, payload);
                    break;
                case 0x8: // close
                    AppLog.Write($"[WS] {conn.RemoteAddr} sent close frame");
                    return;
                case 0x9: // ping → pong
                    await conn.SendRawAsync(0xA, payload);
                    break;
            }
        }
    }

    // Returns (opcode, unmasked payload). Handles fragmented frames.
    private static async Task<(int opcode, byte[] payload)> ReadFrameAsync(NetworkStream stream)
    {
        var fullPayload = new MemoryStream();
        int finalOpcode = 0;

        while (true)
        {
            byte b0 = await ReadByteAsync(stream);
            byte b1 = await ReadByteAsync(stream);
            bool fin    = (b0 & 0x80) != 0;
            int  opcode = b0 & 0x0F;
            bool masked = (b1 & 0x80) != 0;
            long len    = b1 & 0x7F;

            if (len == 126)
            {
                var tmp = await ReadExactAsync(stream, 2);
                len = (tmp[0] << 8) | tmp[1];
            }
            else if (len == 127)
            {
                var tmp = await ReadExactAsync(stream, 8);
                len = 0;
                for (int i = 0; i < 8; i++) len = (len << 8) | tmp[i];
            }

            byte[]? mask = masked ? await ReadExactAsync(stream, 4) : null;
            var payload  = await ReadExactAsync(stream, (int)len);
            if (mask != null)
                for (int i = 0; i < payload.Length; i++)
                    payload[i] ^= mask[i & 3];

            if (opcode != 0) finalOpcode = opcode;
            fullPayload.Write(payload);
            if (fin) return (finalOpcode, fullPayload.ToArray());
        }
    }

    private static async Task<byte> ReadByteAsync(NetworkStream s)
    {
        var b = new byte[1];
        if (await s.ReadAsync(b.AsMemory(0, 1)) != 1) throw new EndOfStreamException();
        return b[0];
    }

    private static async Task<byte[]> ReadExactAsync(NetworkStream s, int count)
    {
        var buf = new byte[count];
        int offset = 0;
        while (offset < count)
        {
            int n = await s.ReadAsync(buf.AsMemory(offset, count - offset));
            if (n == 0) throw new EndOfStreamException();
            offset += n;
        }
        return buf;
    }

    private static async Task<bool> ServeFileAsync(NetworkStream stream, string request)
    {
        try
        {
            var firstLine = request.Split("\r\n")[0];
            var parts     = firstLine.Split(' ');
            if (parts.Length < 2 || parts[0] != "GET") return false;

            var rawPath = parts[1].Split('?')[0];
            var webRoot = Path.Combine(AppContext.BaseDirectory, "web");
            if (!Directory.Exists(webRoot)) return false;

            var rel      = rawPath.TrimStart('/').Replace('/', Path.DirectorySeparatorChar);
            var filePath = string.IsNullOrEmpty(rel)
                ? Path.Combine(webRoot, "index.html")
                : Path.Combine(webRoot, rel);

            if (!File.Exists(filePath))
                filePath = Path.Combine(webRoot, "index.html");
            if (!File.Exists(filePath)) return false;

            var ext  = Path.GetExtension(filePath);
            var mime = GetMime(ext);
            var data = await File.ReadAllBytesAsync(filePath);

            var header = Encoding.ASCII.GetBytes(
                $"HTTP/1.1 200 OK\r\nContent-Type: {mime}\r\nContent-Length: {data.Length}\r\nCache-Control: no-cache\r\nConnection: close\r\n\r\n");
            
            // Send header and data
            await stream.WriteAsync(header, 0, header.Length);
            await stream.WriteAsync(data, 0, data.Length);
            await stream.FlushAsync();
            
            // Wait to ensure data is sent before connection is closed
            await Task.Delay(300);
            
            return true;
        }
        catch { /* ignore broken connections */ }
        return false;
    }

    private static string GetMime(string ext) => ext.ToLowerInvariant() switch
    {
        ".html"  => "text/html; charset=utf-8",
        ".js"    => "application/javascript",
        ".css"   => "text/css",
        ".svg"   => "image/svg+xml",
        ".png"   => "image/png",
        ".ico"   => "image/x-icon",
        ".woff2" => "font/woff2",
        ".woff"  => "font/woff",
        ".json"  => "application/json",
        _        => "application/octet-stream",
    };

    private static string? ExtractHeader(string request, string name)
    {
        var prefix = name + ": ";
        foreach (var line in request.Split("\r\n"))
            if (line.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
                return line[prefix.Length..].Trim();
        return null;
    }
}

sealed class WsConn : IWsConn
{
    private readonly NetworkStream _stream;
    private readonly TcpClient     _tcp;
    private readonly SemaphoreSlim _sendLock = new(1, 1);

    public string RemoteAddr => _tcp.Client.RemoteEndPoint?.ToString() ?? "?";

    public WsConn(NetworkStream stream, TcpClient tcp)
    {
        _stream = stream;
        _tcp    = tcp;
    }

    public async Task SendTextAsync(string text)
    {
        using var frame = BuildFrame(0x1, Encoding.UTF8.GetBytes(text));
        await _sendLock.WaitAsync();
        try
        {
            await _stream.WriteAsync(frame.GetBuffer().AsMemory(0, (int)frame.Length)).ConfigureAwait(false);
        }
        finally { _sendLock.Release(); }
    }

    public async Task SendBinaryAsync(byte[] data)
    {
        using var frame = BuildFrame(0x2, data);
        await _sendLock.WaitAsync();
        try
        {
            await _stream.WriteAsync(frame.GetBuffer().AsMemory(0, (int)frame.Length)).ConfigureAwait(false);
        }
        finally { _sendLock.Release(); }
    }

    public void Disconnect()
    {
        try
        {
            // Send close frame to client
            var closeFrame = new byte[] { 0x88, 0x00 }; // FIN=1, opcode=0x8 (close), length=0
            _stream.Write(closeFrame, 0, closeFrame.Length);
            _stream.Flush();
        }
        catch
        {
            // Ignore errors when sending close frame
        }
        finally
        {
            // Close TCP connection
            _tcp.Close();
        }
    }

    // Sends a WebSocket frame (server → client, no mask, single fragment)
    internal async Task SendRawAsync(byte opcode, byte[] payload)
    {
        using var frame = BuildFrame(opcode, payload);
        await _sendLock.WaitAsync();
        try
        {
            await _stream.WriteAsync(frame.GetBuffer().AsMemory(0, (int)frame.Length)).ConfigureAwait(false);
        }
        finally { _sendLock.Release(); }
    }

    private static MemoryStream BuildFrame(byte opcode, byte[] payload)
    {
        var ms = new MemoryStream(10 + payload.Length);
        ms.WriteByte((byte)(0x80 | opcode)); // FIN=1
        if (payload.Length < 126)
        {
            ms.WriteByte((byte)payload.Length);
        }
        else if (payload.Length < 65536)
        {
            ms.WriteByte(126);
            ms.WriteByte((byte)(payload.Length >> 8));
            ms.WriteByte((byte)(payload.Length));
        }
        else
        {
            ms.WriteByte(127);
            long len64 = payload.Length;
            for (int i = 7; i >= 0; i--)
                ms.WriteByte((byte)(len64 >> (i * 8)));
        }
        ms.Write(payload);
        return ms;
    }
}
